atomicfu-transformer/src/main/kotlin/kotlinx/atomicfu/transformer/AtomicFUTransformer.kt

```diff
@@ -785,46 +775,126 @@ class AtomicFUTransformer(
+                    } else {
+                        instructions.remove(ld)
+                    }
+                    val lv = localVar(v, operation)
                     if (lv != null) {
                         // Stored to a local variable with an entry in LVT (typically because of inline function)
-                        if (lv.desc != f.fieldType.descriptor)
+                        if (lv.desc != f.fieldType.descriptor && !onArrayElement)
                             abort("field $f was stored to a local variable #$v \"${lv.name}\" with unexpected type: ${lv.desc}")
                         // correct local variable descriptor
                         lv.desc = f.ownerType.descriptor
                         lv.signature = null
                         // process all loads of this variable in the corresponding local variable range
                         forVarLoads(v, lv.start, lv.end) { otherLd ->
-                            fixupVarLoad(f, ld, otherLd)
+                            fixupLoad(f, ld, otherLd, arrayElementInfo)
                         }
                     } else {
                         // Spilled temporarily to a local variable w/o an entry in LVT -> fixup only one load
-                        fixupVarLoad(f, ld, nextVarLoad(v, next))
+                        fixupLoad(f, ld, nextVarLoad(v, next), arrayElementInfo)
                     }
                     return next
                 }
                 else -> abort("cannot happen")
             }
         }
 
+        private fun fixupLoad(f: FieldInfo, ld: FieldInsnNode, otherLd: VarInsnNode, arrayElementInfo: List<AbstractInsnNode>?): AbstractInsnNode? {
+            return if (arrayElementInfo != null) {
+                fixupArrayElementLoad(f, ld, otherLd, arrayElementInfo)
+            } else {
+                fixupVarLoad(f, ld, otherLd)
+            }
+        }
+
         private fun fixupVarLoad(f: FieldInfo, ld: FieldInsnNode, otherLd: VarInsnNode): AbstractInsnNode? {
             val ldCopy = ld.clone(null) as FieldInsnNode
             instructions.insert(otherLd, ldCopy)
             return fixupLoadedAtomicVar(f, ldCopy)
         }
 
+        private fun fixupArrayElementLoad(f: FieldInfo, ld: FieldInsnNode, otherLd: VarInsnNode, arrayElementInfo: List<AbstractInsnNode>): AbstractInsnNode? {
+            // index instructions from array element info: drop owner class load instruction (in vh case together with preceding getting VH + swap)
+            val index = arrayElementInfo.drop(if (vh) 3 else 1)
+            // previously stored array element reference is loaded -> arrayElementInfo should be cloned and inserted at the point of this load
+            // before cloning make sure that index instructions contain just loads and simple arithmetic, without any invocations and complex data flow
+            for (indexInsn in index) {
+                checkDataFlowComplexity(indexInsn)
+            }
```

Comment:
Good. Can we just do this verification loop in `fixupLoadedArrayElement` right where this `arrayElementInfo` is created?